test/test_intensify/test_parallel_successive_halving.py

```diff
@@ -0,0 +1,391 @@
+        # This means we run int(2.5) + 2.0 = 4 runs before waiting
+        # But we have 2 successive halvers now!
+        self.assertEqual(i, 8)
+
+    def test_add_new_SH(self):
+        """Test whether we can add a SH and when we should not"""
+
+        # Case of a single SH
+        self.PSH.max_active_SH = 1
+
+        # By default we create at least 1 SH
+        self.assertEqual(len(self.PSH.sh_instances), 1)
+        self.assertIsInstance(self.PSH.sh_instances[0], SuccessiveHalving)
+
+        # If max active SH is called, we should not add another SH
+        # and return false
+        self.assertFalse(self.PSH._add_new_SH())
+        self.assertEqual(len(self.PSH.sh_instances), 1)
+
+        # We try with 2 SH active
+        self.PSH.max_active_SH = 2
+
+        # We effectively return true because we added a new SH
+        self.assertTrue(self.PSH._add_new_SH())
+
+        self.assertEqual(len(self.PSH.sh_instances), 2)
+        self.assertIsInstance(self.PSH.sh_instances[1], SuccessiveHalving)
+
+        # Trying to add a third one should return false
+        self.assertFalse(self.PSH._add_new_SH())
+        self.assertEqual(len(self.PSH.sh_instances), 2)
+
+    def test_sh_scheduling(self):
+        """Ensures proper scheduling is returned based on the strategy"""
+
+        # Add more SH to make testing interesting
+        self.PSH.max_active_SH = 4
+        for i in range(4):
+            self.PSH._add_new_SH()
+
+        # First test the ordinal approach
+        self.assertCountEqual(list(self.PSH._sh_scheduling(strategy='ordinal')), [0, 1, 2, 3])
+
+        # Also, if strategy is serial and index is 3 it should be the same as ordinal
+        # That is, we lastly queried the SH #3, we have to start from 0 again
+        self.PSH.last_active_instance = 3
+        self.assertCountEqual(list(self.PSH._sh_scheduling(strategy='serial')), [0, 1, 2, 3])
+
+        # Make sure we complete the list properly in serial
+        self.PSH.last_active_instance = 0
+        self.assertCountEqual(list(self.PSH._sh_scheduling(strategy='serial')), [1, 2, 3, 0])
+        self.PSH.last_active_instance = 1
+        self.assertCountEqual(list(self.PSH._sh_scheduling(strategy='serial')), [2, 3, 0, 1])
+        self.PSH.last_active_instance = 2
+        self.assertCountEqual(list(self.PSH._sh_scheduling(strategy='serial')), [3, 0, 1, 2])
+
+    def _exhaust_run_and_get_incumbent(self, sh, rh):
+        """
+        Runs all provided configs on all instances and return the incumbent
+        as a nice side effect runhistory/stats are properly filled
+        """
+        challengers = [self.config1, self.config2, self.config3, self.config4]
+        incumbent = None
+        for i in range(100):
+            try:
+                intent, run_info = sh.get_next_run(
+                    challengers=challengers,
+                    incumbent=None, chooser=None, run_history=rh)
+            except ValueError as e:
```

Comment:
Running out of configs shouldn't raise a ValueError as this is not about calling something with an illegal value, isn't it? Is there a more appropriate exception we could raise?

Reply:
This is a testing only problem, because we really do not provide an EPM:

https://github.com/automl/SMAC3/blob/9890e0cdc4ce898de25fa00d794cbe0af83a09c2/smac/intensification/abstract_racer.py#L222

It makes sense as a valueerror because there needs to be epm or enough configs from where to sample.

Do you think this should be changed? In real execution,smac creates the EPM so this is unlikely to happen

Reply:
Ah, okay, that makes sense then.